src/main/scala/io/iohk/ethereum/network/p2p/Message.scala

```diff
@@ -60,44 +64,427 @@ object Hello {
+
+      case _ => throw new RuntimeException("Cannot decode GetBlockHeaders")
+    }
+  }
+
+  val code: Int = 0x10 + 0x03
+}
+
+case class GetBlockHeaders(block: Either[BigInt, ByteString], maxHeaders: BigInt, skip: BigInt, reverse: Int) extends Message {
+  override def code: Int = GetBlockHeaders.code
+
+  override def toString: String = {
+    s"""GetBlockHeaders{
+       |block: ${block.fold(a => a, b => Hex.toHexString(b.toArray[Byte]))}
+       |maxHeaders: $maxHeaders
+       |skip: $skip
+       |reverse: ${reverse == 1}
+       |}
+     """.stripMargin
+  }
+}
+
+object BlockHeaders {
+  implicit val rlpEndDec = new RLPEncoder[BlockHeaders] with RLPDecoder[BlockHeaders] {
+    override def encode(obj: BlockHeaders): RLPEncodeable = {
+      import obj._
+      RLPList(headers.map(BlockHeader.rlpEndDec.encode): _*)
+    }
+
+    override def decode(rlp: RLPEncodeable): BlockHeaders = rlp match {
+      case rlpList: RLPList => BlockHeaders(rlpList.items.map(BlockHeader.rlpEndDec.decode))
+
+      case _ => throw new RuntimeException("Cannot decode BlockHeaders")
+    }
+  }
+
+  val code: Int = 0x10 + 0x04
+}
+
+case class BlockHeaders(headers: Seq[BlockHeader]) extends Message {
+  override def code: Int = BlockHeaders.code
+}
+
+object BlockHeader {
+  implicit val rlpEndDec = new RLPEncoder[BlockHeader] with RLPDecoder[BlockHeader] {
+    override def encode(obj: BlockHeader): RLPEncodeable = {
+      import obj._
+      RLPList(
+        parentHash.toArray[Byte],
+        ommersHash.toArray[Byte],
+        beneficiary.toArray[Byte],
+        stateRoot.toArray[Byte],
+        transactionsRoot.toArray[Byte],
+        receiptsRoot.toArray[Byte],
+        logsBloom.toArray[Byte],
+        difficulty,
+        number,
+        gasLimit,
+        gasUsed,
+        unixTimestamp,
+        extraData.toArray[Byte],
+        mixHash.toArray[Byte],
+        nonce.toArray[Byte])
+    }
+
+    override def decode(rlp: RLPEncodeable): BlockHeader = rlp match {
+      case RLPList(parentHash, ommersHash, beneficiary, stateRoot, transactionsRoot, receiptsRoot, logsBloom,
+      difficulty, number, gasLimit, gasUsed, unixTimestamp, extraData, mixHash, nonce) =>
+        BlockHeader(ByteString(parentHash: Array[Byte]),
+          ByteString(ommersHash: Array[Byte]),
+          ByteString(beneficiary: Array[Byte]),
+          ByteString(stateRoot: Array[Byte]),
+          ByteString(transactionsRoot: Array[Byte]),
+          ByteString(receiptsRoot: Array[Byte]),
+          ByteString(logsBloom: Array[Byte]),
+          difficulty,
+          number,
+          gasLimit,
+          gasUsed,
+          unixTimestamp,
+          ByteString(extraData: Array[Byte]),
+          ByteString(mixHash: Array[Byte]),
+          ByteString(nonce: Array[Byte]))
+
+      case _ => throw new RuntimeException("Cannot decode BlockHeaders")
+    }
+  }
+}
+
+case class BlockHeader(
+  parentHash: ByteString,
```

Comment:
as far as I remember the guide suggests to use 2 indentation levels here, i.e 4 spaces